src/validation.cpp

```diff
@@ -1839,20 +1838,16 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         if (!tx.IsCoinBase())
         {
             CAmount txfee = 0;
-            if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, txfee)) {
-                if (!IsBlockReason(state.GetReason())) {
```

Comment:
In commit "[validation] Add CValidationState subclasses" (b31987c45045cc26ca7867d9d8aa645f5d36b5e6)

To be sure, is there no change in behavior here because IsBlockReason would always return false if CheckTxInputs set failed with any reason other than CONSENSUS? Or is there no change in behavior for some other reason? Or is there a actually a change in some cases?

Related question: Is it a problem going forward that TxValidationResult::RECENT_CONSENSUS_CHANGE will be translated to BlockValidationResult::CONSENSUS_CHANGE instead of BlockValidationResult::RECENT_CONSENSUS_CHANGE?

Reply:
Take a look at `CheckTxInputs()`. Prior to this commit, if it returned false, then `state.m_reason` could be set to:

- `TX_MISSING_INPUTS`
- `TX_PREMATURE_SPEND`
- `CONSENSUS`

In all cases, that would be converted to `CONSENSUS`, so this is not a change in behaviour.

> Is it a problem going forward that TxValidationResult::RECENT_CONSENSUS_CHANGE will be translated to BlockValidationResult::CONSENSUS_CHANGE instead of BlockValidationResult::RECENT_CONSENSUS_CHANGE?

I don't think so. Anyone introducing `RECENT_CONSENSUS_CHANGE` should verify that it gets properly propagated up the call stack.